daemon/vpn/quench/libquench.go

```diff
@@ -0,0 +1,272 @@
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	opts := quenchBindigns.NewVnicOptions()
+	opts.SetFwmark(q.fwmark)
+
+	vnic, err := quenchBindigns.VnicFromName(q.vnicName, q.observer, &opts)
+	if err != nil {
+		return fmt.Errorf("creating vnic instance: %w", err)
+	}
+
+	iface, err := net.InterfaceByName(q.vnicName)
+	if err != nil {
+		return fmt.Errorf("retrieving the interface: %w", err)
+	}
+
+	ip := netip.MustParsePrefix(quenchInterfaceAddr)
+	tun := tunnel.New(*iface, []netip.Addr{}, ip)
+
+	if err := tun.AddAddrs(); err != nil {
+		return fmt.Errorf("setting up vinc: %w", err)
+	}
+
+	if err := tun.Up(); err != nil {
+		return fmt.Errorf("adding ip address to vnic: %w", err)
+	}
+
+	addr := fmt.Sprintf("wt://%s:%d/", server.IP, server.NordWhisperPort)
+
+	config := Config{
+		Protocol: Protocol{
+			Addr: addr,
+			Spec: Spec{
+				TlsDomain: server.Hostname,
+			},
+		},
+	}
+
+	jsonConfig, err := json.Marshal(config)
+	if err != nil {
+		return fmt.Errorf("marshaling json config: %w", err)
+	}
+
+	log.Println(internal.DebugPrefix, "quench config:", string(jsonConfig))
+
+	quenchCreds := quenchBindigns.Credentials{
+		User: creds.OpenVPNUsername,
+		Pass: creds.OpenVPNPassword,
+	}
+
+	eventsContext, eventsCancelFunc := context.WithCancel(ctx)
+	eventsChan := q.observer.SubscribeToEvents(eventsContext)
+	defer eventsCancelFunc()
+
+	err = vnic.Connect(string(jsonConfig), &quenchCreds)
+	if err != nil {
+		return fmt.Errorf("connecting to a quench server: %w", err)
+	}
+
+	q.vnic = vnic
+	q.tun = tun
+	q.server = server
+
+	log.Println(internal.DebugPrefix, "waiting for connection")
+CONNECTION_LOOP:
+	for {
+		select {
+		case <-ctx.Done():
+			log.Println(internal.DebugPrefix, "context cancelled before connection was established")
+			return ctx.Err()
+		case ev := <-eventsChan:
+			q.state = ev
+
+			if ev == vpn.ExitedState {
+				return fmt.Errorf("connection failed")
+			}
+
+			if ev == vpn.ConnectedState {
+				break CONNECTION_LOOP
+			}
+		}
+	}
+
+	return nil
+}
+
+func (q *Quench) Stop() error {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	eventsChan := q.observer.SubscribeToEvents(ctx)
+	defer cancelFunc()
+
+	q.state = vpn.ExitingState
+
+	if q.vnic != nil {
+		if err := q.vnic.Disconnect(); err != nil {
+			return fmt.Errorf("disconnectiong from a quench server: %w", err)
+		}
+
+		for {
+			ev := <-eventsChan
+			if ev == vpn.ExitedState {
+				break
+			}
+		}
+
+		q.vnic.Destroy()
+		q.vnic = nil
+	}
+
+	q.tun = nil
+	q.state = vpn.ExitedState
+
+	return nil
+}
+
+func (q *Quench) State() vpn.State {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	return q.state
+}
+
+func (q *Quench) IsActive() bool {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.state == vpn.ConnectedState || q.state == vpn.ConnectingState
+}
+
+func (q *Quench) Tun() tunnel.T {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.tun
```

Comment:
I think you'll need to check here if `tun` is `nil` and return `nil` in that case otherwise the return will never be `nil`